File: Week 5/2. List order of nodes visited by a DFS/dfs_list_order_nodes.cpp
#include <iostream>
#include <vector>

using namespace std;

class Graph
{
private:
    int numVertices;
    vector<vector<int>> adj;
    vector<bool> visited;

public:
    Graph(int numVertices)
    {
        this->numVertices = numVertices;
        adj.resize(numVertices);
        visited.resize(numVertices, false);
    }

    void add_edge(int u, int v)
    {
        adj[u - 1].push_back(v - 1);
        adj[v - 1].push_back(u - 1);
    }

    void dfs(int u)
    {
        visited[u] = true;
        cout << u + 1 << " ";

        for (int v : adj[u])
            if (!visited[v])
                dfs(v);
    }
};

int main()
{
    int n, m;
    cin >> n >> m;

    Graph g(n);

    for (int i = 0; i < m; ++i)
    {
        int u, v;
        cin >> u >> v;
        g.add_edge(u, v);
    }

    g.dfs(0);
}

File: Week 5/2. List order of nodes visited by a DFS/dfs_list_order_nodes.py
def add_edge(adj, u, v):
    adj[u - 1].append(v - 1)
    adj[v - 1].append(u - 1)
    
def dfs_rec(adj, visited, u):
    visited[u] = True
    
    print(u + 1, end = " ")
    
    for i in adj[u]:
        if not visited[i]:
            dfs_rec(adj, visited, i)

def dfs(adj, u):
    visited = [False] * len(adj)
    dfs_rec(adj, visited, u)
    
if __name__ == "__main__":
    n, m = map(int, input().split())
    adj = [[] for _ in range(n)]
    
    for _ in range(m):
        u, v = map(int, input().split())
        add_edge(adj, u, v)
        
    dfs(adj, 0)
